### PR TITLE
fix: allow new booking to overlap old one when rescheduling

### LINKED ISSUE #2668: [CAL-1685] "Edit booking" doesn't allow a new time which overlaps old time
### Issue Summary

Someone booked a 60 minute slot via my cal.com event link, but then pointed out they wanted to move it 15 minutes later; however the "edit booking" feature doesn't allow this.

### Steps to Reproduce

0. Confirm a 60 minute booking using an event type which has `Time-slot intervals` set to 15 minutes.  (In my case, the booking was at 11:30am.)
1. Go to https://app.cal.com/bookings/upcoming
2. Click on "Reschedule"
3. Click on "Edit booking"
4. Select the same day as the original booking time
5. Observe that the times available for rebooking don't overlap with the original time, so I can't shift the original 60 minute booking at 11:30am earlier or later by 15 or 30 or 45 minutes:

image

### Expected behaviour

When rescheduling, the availability algorithm should consider the existing booking as (provisionally) cancelled just for the purposes of editing its time, in order to avoid the old time incorrectly overlapping with potential new times.

### Technical details

Google Chrome Version 101.0.4951.41 (Official Build) (64-bit)

This is clearly a server-side issue, but happy to provide more client details if someone needs them.

<sub>[CAL-1685](https://linear.app/calcom/issue/CAL-1685/edit-booking-doesnt-allow-a-new-time-which-overlaps-old-time)</sub>

### PR DESCRIPTION
**Pull request by CarinaWolli**
## What does this PR do?

When rescheduling we now also allow to book slots that were blocked from the original booking. 

Fixes #2668

Loom Video: https://www.loom.com/share/b782bcdb97894bf183ed526526e28240

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How should this be tested?

Test Steps to reproduce as described in linked issue: 

- Book a 60 minute booking using an event type which has Time-slot intervals set to 15 minutes. (For example at 11:30am)
- Reschedule the booking
- Select the same day as the original booking time
- See that now all slots from the original booking are available

Test with and without a connected calendar.

## Checklist

- [x] add test

### CODE REVIEW

File: packages/core/getBusyTimes.ts [hunk 1]
```diff
@@ -69,9 +73,14 @@ export async function getBusyTimes(params: {
    */
   performance.mark("prismaBookingGetStart");
 
+  const startTimeDate =
+    rescheduleUid && duration ? dayjs(startTime).subtract(duration, "minute").toDate() : new Date(startTime);
```
Comment (CarinaWolli) [inline packages/core/getBusyTimes.ts:77]: On the time of booking a slot (handleNewBooking) `startTime` is the start time of the new booking, which means that the old booking would not be included in `bookings`. If we add the duration of the original booking it should always be included when querying `bookings`

File: packages/core/getBusyTimes.ts [hunk 2]
```diff
@@ -140,6 +150,9 @@ export async function getBusyTimes(params: {
         // doing this allows using the map later to remove the ranges from calendar busy times.
         delete bookingSeatCountMap[bookedAt];
       }
+      if (rest.uid === rescheduleUid) {
+        return aggregate;
```
Comment (CarinaWolli) [inline packages/core/getBusyTimes.ts:154]: return early, as this is the original rescheduled booking and should not be included in busy times

File: packages/core/getBusyTimes.ts [hunk 3]
```diff
@@ -183,6 +196,17 @@ export async function getBusyTimes(params: {
       };
     });
 
+    if (rescheduleUid) {
+      const originalRescheduleBooking = bookings.find((booking) => booking.uid === rescheduleUid);
+      // calendar busy time from original rescheduled booking should not be blocked
+      if (originalRescheduleBooking) {
+        openSeatsDateRanges.push({
```
Comment (CarinaWolli) [inline packages/core/getBusyTimes.ts:203]: `openSeatsDateRanges` is be substracted from the `calendarBusyTimes` later

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 4]
```diff
@@ -851,6 +865,34 @@ async function handler(
...
+    }
+    originalRescheduledBooking = await getOriginalRescheduledBooking(
+      rescheduleUid,
+      !!eventType.seatsPerTimeSlot
+    );
+    if (!originalRescheduledBooking) {
+      throw new HttpError({ statusCode: 404, message: "Could not find original booking" });
+    }
+  }
+
```
Comment (CarinaWolli) [inline packages/features/bookings/lib/handleNewBooking.ts:895]: just moved up here because I need `orginalRescheduledBooking`, no code changes

**Review by Udit-takkar — COMMENTED** [review]
Screenshot 2023-07-28 at 4 03 00 PM

I first scheduled for 11:00 AM ( 60 minute event, 15 minute interval slots). I don't see any slots between 10:00AM and 12:00PM while rescheduling

**Comment by CarinaWolli** [comment]
> I first scheduled for 11:00 AM ( 60 minute event, 15 minute interval slots). I don't see any slots between 10:00AM and 12:00PM while rescheduling

@Udit-takkar With calendar connected or without?

File: packages/core/getBusyTimes.ts [hunk 5]
```diff
@@ -183,6 +196,17 @@ export async function getBusyTimes(params: {
       };
     });
 
+    if (rescheduleUid) {
```
Comment (CarinaWolli) [inline packages/core/getBusyTimes.ts:199]: @Udit-takkar I could reproduce the issue you mentioned, however not always and it is working most of the time for me. I couldn't figure out yet why it sometimes isn't working but I think when it fails the rescheduleUid isn't coming in

**Comment by CarinaWolli** [comment]
@Udit-takkar made some fixes, can you re-review?

File: apps/web/playwright/reschedule.e2e.ts [hunk 6]
```diff
@@ -239,4 +240,27 @@ test.describe("Reschedule Tests", async () => {
...
+    const eventType = user.eventTypes[0];
+    const startTime = dayjs().add(1, "day").set("hour", 10).set("minute", 0).toDate();
+    const endTime = dayjs().add(1, "day").set("hour", 10).set("minute", 30).toDate();
+
+    const booking = await bookings.create(user.id, user.username, eventType.id, {}, startTime, endTime);
+
+    await page.goto(`/reschedule/${booking.uid}`);
+
+    //book same slot again
+    page.getByRole("button", { name: dayjs(startTime).format("D") }).click();
```
Comment (zomars) [inline apps/web/playwright/reschedule.e2e.ts:259]: Seems like we're getting 5 buttons with the same startTime

image

File: packages/features/bookings/Booker/components/DatePicker.tsx [hunk 7]
```diff
@@ -10,15 +10,15 @@ import { useLocale } from "@calcom/lib/hooks/useLocale";
...
+export const DatePicker = ({ rescheduleUid }: { rescheduleUid?: string | null }) => {
   const { i18n } = useLocale();
   const [month, selectedDate] = useBookerStore((state) => [state.month, state.selectedDate], shallow);
   const [setSelectedDate, setMonth] = useBookerStore(
     (state) => [state.setSelectedDate, state.setMonth],
     shallow
   );
   const event = useEvent();
-  const schedule = useScheduleForEvent();
+  const schedule = useScheduleForEvent({ rescheduleUid });
```
Comment (zomars) [inline packages/features/bookings/Booker/components/DatePicker.tsx:21]: Can we get rescheduleUid from the URL instead of drilling props down?

**Review by zomars — DISMISSED** [review]
Thank you @CarinaWolli Left some observations about prop-drilling and e2e tests failing.

File: apps/web/playwright/reschedule.e2e.ts (hunk 6 above)
Comment (CarinaWolli) [reply apps/web/playwright/reschedule.e2e.ts]: fixed 👍

**Review by Udit-takkar — CHANGES_REQUESTED** [review]
Screenshot 2023-08-08 at 1 29 14 PM

Tried again with no connected calednar but i still don't see 11:15, 11:30, 11:45

Edit: I see it now

Screenshot 2023-08-08 at 1 43 42 PM

**Review by Udit-takkar — DISMISSED** [review]
LGTM. works fine now

**Comment by CarinaWolli** [comment]
@Udit-takkar lost your approval when solving merge conflicts, could you please re-approve 🙏

**Review by zomars — CHANGES_REQUESTED** [review]
Can we check why the test are failing in this one? 🙏🏽

File: packages/features/bookings/lib/handleNewBooking.ts [hunk 8]
```diff
@@ -344,22 +345,35 @@ async function ensureAvailableUsers(
...
         userId: user.id,
         eventTypeId: eventType.id,
+        duration: orginalBookingDuration,
         ...input,
       },
-      { user, eventType }
+      {
+        user,
+        eventType,
+        rescheduleUid: input.originalRescheduledBooking ? input.originalRescheduledBooking.uid : null,
```
Comment (alannnc) [inline packages/features/bookings/lib/handleNewBooking.ts:375]:
```suggestion
        rescheduleUid: input.originalRescheduledBooking?.uid ?? null,
```

**Review by alannnc — DISMISSED** [review]
Looks good and have a test for this scenario! 👍

**Comment by alannnc** [comment]
Looks like seats test failed, let me help you with that.

File: packages/features/bookings/Booker/utils/event.ts [hunk 9]
```diff
@@ -58,13 +59,16 @@ export const useScheduleForEvent = ({
     (state) => [state.username, state.eventSlug, state.month, state.selectedDuration],
     shallow
   );
+  const serachParams = useSearchParams();
+  const rescheduleUid = serachParams.get("rescheduleUid");
```
Comment (zomars) [inline packages/features/bookings/Booker/utils/event.ts:63]: Updated to use next new hooks for consistency. Type signature remains the same.

**Review by zomars — APPROVED** [review]
[nice](https://app.graphite.dev/macros/calcom)

**Review by shivamklr — COMMENTED** [review]
@Udit-takkar @CarinaWolli This test started to fail recently. Not really sure what exactly is wrong here.

**Comment by CarinaWolli** [comment]
I see it failing here: https://github.com/calcom/cal.com/pull/10606. I will take a look. Thank you @shivamklr

**Pull request merged**